in_toto/model.go

```diff
@@ -150,6 +190,67 @@ func (l *Layout) InspectAsInterfaceSlice() []interface{} {
 	return inspectionsI
 }
 
+func (l *Layout) validate() {
+	validateFunctions := []func(){l.validateType, l.validateExpires,
+		l.validateKeys, l.validateStepsAndInspections}
+	for _, f := range validateFunctions {
+		f()
+	}
+}
+
+func (l *Layout) validateType() {
+	if l.Type != "layout" {
+		fmt.Println("Invalid Type value for layout: should be 'layout'")
+	}
+}
+
+func (l *Layout) validateExpires() {
+	if _, err := time.Parse("2006-01-02T15:04:05Z", l.Expires); err != nil {
+		fmt.Println("Expiry time parsed incorrectly - date either invalid or" +
+			" of incorrect format")
+	}
+}
+
+func (l *Layout) validateKeys() {
+	for keyId, key := range l.Keys {
```

Comment:
If we had a `Key` validator, we could use it here and elsewhere. What do you think? 
So instead of below code you just do

```go
for keyId, key := range l.Keys {
  key.validate()
...
```

Reply:
Good call.

Reply:
Do you have an idea for how to distinguish between public and private keys? `key.validate` must check that the private part is empty for a public key, and I was wondering if that should just be a separate check? If it is, it doesn't really leave enough validation processes to warrant being a part of the `Key` struct I think.

Reply:
Not sure, I understand what you mean by your last sentence. Here are some thoughts:

In the reference implementation we use an OOP-approach for the top level metadata, e.g. if you call `layout.validate()` on a `Layout` object, it knows how to validate itself.

For lower level metadata we use securesystemslib's approach, which although implemented with OOP is actually functional in regard to the validated data, i.e. if you want to validate a key, it does not know what kind of key it is or that it is a key at all, so you must specify as what you want to validate it, e.g. `PUBLIC_KEY_SCHEMA.matches(something)`.

AFAICS go lends itself towards the second approach, because it has very stripped-down OOP concepts (e.g. no inheritance). So to get back to your original question, I'd probably go for multiple key validation functions, along the lines `validateAnyKey`, `validatePubKey`, `validatePrivKey`.

It might also be worth re-considering the general use of receivers, e.g. `key.validateAsPubKey` vs. `validatePubKey(key)`. 

Does this make sense? What do you think?

Reply:
Yes, that makes sense. Since you'd mentioned looking into validating it as `key.validate()` and letting it take care of itself, I was thinking about how to do the check for it being a private key. However, I'll do something like this instead: https://github.com/adityasaky/in-toto-golang/blob/add-validate/in_toto/model.go#L36 which we can reuse if necessary. This would also mean that the `Key` struct has no validation methods itself (using a receiver).